### ISSUE TITLE
diffmax_t is incompatible with gcc std::ranges

### ISSUE THREAD

**Issue by patlefort** [ericniebler/range-v3#1516]
diffmax_t will not work with some std::ranges stuff as it is not recognized as a signed integer. For example, this will not compile, on a 64 bits system:
```cpp
std::ranges::all_of(ranges::views::ints(std::uint64_t{}, std::uint64_t{10}), [](auto i){ return i < 10; });
```
In `bits/iterator_concepts.h`:
```cpp
template<typename _Tp>
concept __is_signed_integer_like = signed_integral<_Tp>
	|| same_as<_Tp, __max_diff_type>;
```
The chain of concept is: input_range -> input_iterator -> input_or_output_iterator -> weakly_incrementable -> __is_signed_integer_like.

I am not sure what is the best solution, but a possible solution would be to use __int128, which is what libstdc++ is using, if it's available instead of the current diffmax_t struct.

**Comment by marehr** [comment]
Duplicate of https://github.com/ericniebler/range-v3/issues/1514?

**Comment by patlefort** [comment]
It's tangentially related. If I understand correctly, his problem is that std ranges might not be considered a valid range in rangev3 if they use __int128, my problem is that rangev3 ranges aren't considered valid ranges in std ranges if they use diffmax_t. The root problem is the different int type over 64 bits that make ranges incompatible with one another.

**Comment by marehr** [comment]
You are right, the problem is symmetric, I tried to fix only one direction (valid `std::views::view` -> valid `::ranges::view`).

`::ranges::range_difference_t<decltype(ranges::views::ints(std::uint64_t{}, std::uint64_t{10}))>` is `ranges::detail::diffmax_t` which returns `std::is_signed_v == false` and will cause `std::indirectly_readable` to fail.

I guess we need a specialisation for `std::is_signed<::ranges::detail::diffmax_t>` like we have for `numeric_limits` or should use `__int128` directly if available.

https://github.com/ericniebler/range-v3/blob/f5af689d49457a6c625854182bc20725b6cfc6be/include/range/v3/iterator/diffmax_t.hpp#L359-L365

**Comment by johelegp** [comment]
`is_signed` can't be specialized, like most standard traits.

**Comment by marehr** [comment]
Hmm, I thought `numeric_limits` is off-limits to, but https://en.cppreference.com/w/cpp/types/numeric_limits says that 
"Implementations may provide specializations of std::numeric_limits for implementation-specific types". I couldn't find that passage in the standard tough.

This means the standard could have extended the "concept" of [`template<class T> inline constexpr bool is-signed-integer-like`](https://eel.is/c++draft/iterator.concept.winc#1) by `= true` iff `std::numeric_limits<T>::is_signed` and everything would have work for custom integer types?

**Comment by johelegp** [comment]
It is. See https://timsong-cpp.github.io/cppwp/n4861/numeric.limits#6.

**Comment by patlefort** [comment]
The standard says that "Non-arithmetic standard types" shall not have specializations, not that it can't be specialized by your own, non-standard types.

**Comment by johelegp** [comment]
You're absolutely right.

**Comment by patlefort** [comment]
To add to the problems with diffmax_t, it also doesn't work with OpenMP as it is not recognized as a valid integer, so parallel range for loops will not work.
